File: app/src/main/java/org/ametro/ui/controllers/MotionEventWrapper.java
package org.ametro.ui.controllers;

import android.util.Log;
import android.view.MotionEvent;

public class MotionEventWrapper {

	public static final String TAG = "MotionEventWrapper";
	
	public static final int ACTION_POINTER_UP;
	public static final int ACTION_POINTER_DOWN;
	
	public static final boolean HasMultiTouchSupport;

	protected MotionEvent event;
	
	static {
		boolean succeeded = false;
		try {
			MotionEvent.class.getMethod("getPointerCount");
			succeeded = true;
		} catch (Exception e) {
			Log.e(TAG, "Methods static initializer failed", e);
		}
		HasMultiTouchSupport = succeeded;
		int pointerDown = 5;
		int pointerUp = 6;
		if (HasMultiTouchSupport) {
			try{
				pointerDown = MotionEvent.class.getField("ACTION_POINTER_DOWN").getInt(null);
				pointerUp = MotionEvent.class.getField("ACTION_POINTER_UP").getInt(null);
			}catch(Exception e){
				Log.e(TAG, "Constants static initializer failed", e);
			}
		}
		ACTION_POINTER_DOWN = pointerDown;
		ACTION_POINTER_UP = pointerUp;
	}	
	
	public MotionEventWrapper(MotionEvent event) {
		this.event = event;
	}

	public static MotionEventWrapper create(MotionEvent event){
		if(HasMultiTouchSupport){
			return new MotionEventMultiTouchWrapper(event);
		}else{
			return new MotionEventWrapper(event);
		}
	}
	
	public int getAction(){
		return event.getAction();
	}
	
	public float getX(){
		return event.getX();
	}
	
	public float getY(){
		return event.getY();
	}
	
	public float getX(int pos){
		return 0; 
	}
	
	public float getY(int pos){
		return 0;
	}

	public MotionEvent getEvent() {
		return event;
	}

	public long getEventTime() {
		return event.getEventTime();
	}
	
	
	
}

File: app/src/main/java/io/github/romangolovanov/apps/ametro/ui/controllers/MotionEventWrapper.kt
package io.github.romangolovanov.apps.ametro.ui.controllers

import android.util.Log
import android.view.MotionEvent

open class MotionEventWrapper(internal var event: MotionEvent) {

    companion object {
        const val TAG = "MotionEventWrapper"

        val ACTION_POINTER_UP: Int
        val ACTION_POINTER_DOWN: Int
        val HasMultiTouchSupport: Boolean

        init {
            var succeeded = false
            try {
                MotionEvent::class.java.getMethod("getPointerCount")
                succeeded = true
            } catch (e: Exception) {
                Log.e(TAG, "Methods static initializer failed", e)
            }
            HasMultiTouchSupport = succeeded
            var pointerDown = 5
            var pointerUp = 6
            if (HasMultiTouchSupport) {
                try {
                    pointerDown = MotionEvent::class.java.getField("ACTION_POINTER_DOWN").getInt(null)
                    pointerUp = MotionEvent::class.java.getField("ACTION_POINTER_UP").getInt(null)
                } catch (e: Exception) {
                    Log.e(TAG, "Constants static initializer failed", e)
                }
            }
            ACTION_POINTER_DOWN = pointerDown
            ACTION_POINTER_UP = pointerUp
        }

        fun create(event: MotionEvent): MotionEventWrapper {
            return if (HasMultiTouchSupport) {
                MotionEventMultiTouchWrapper(event)
            } else {
                MotionEventWrapper(event)
            }
        }
    }

    open fun getAction(): Int = event.action

    open fun getX(): Float = event.x

    open fun getY(): Float = event.y

    open fun getX(pos: Int): Float = 0f

    open fun getY(pos: Int): Float = 0f

    fun getEventTime(): Long = event.eventTime
}
